src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        // We attempt to make error recovery a little nicer here by only treating a single
+        // identifier token as a calling convention if the token afterwords is actually a
+        // possible start to the function pointer parameter list. The intuition is that, for
+        // some partial trees like this:
+        //
+        //  (delegate*MyProperty.MyMethod();
+        //
+        // The user is actually in the middle of adding a cast, and if we interpreted MyProperty
+        // as a calling convention, we'd get a much worse set of errors than if we treated the
+        // function pointer as having no calling convention.
+        // Also for nicer error recovery, we only treat predefined types as if they could be the
+        // calling convention. For any other keyword, they're almost certainly part of some other
+        // construct, and would produce better errors if parsed separately. The user could have this:
+        //
+        // delegate* // Declaration in progress
+        // while (true) {}
+        //
+        // The parse tree will look much better if the while(true) is considered a separate structure,
+        // even though it does look like it could be the start of an invalid function pointer definition.
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
+               // Treat predefined types as possible calling conventions, as they're a single
+               // token. This will be a semantic error, but it'll result in a nicer parse tree.
+               (CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind));
```

Comment:
>IsPredefinedType [](start = 67, length = 16)

Is there a reason to limit this to just predefined types? Would we want to allow any single "word" (```IsSomeWord```)?

Reply:
This is covered in the comment before the function

---
In reply to: [349358002](https://github.com/dotnet/roslyn/pull/39879#discussion_r349358002) [](ancestors = 349358002)